compiler/rustc_mir_build/src/thir/pattern/check_match.rs

```diff
@@ -412,27 +394,33 @@ fn irrefutable_let_pattern(id: HirId, ls: LetSource, span: Span, tcx: TyCtxt<'_>
         }};
     }
 
+    let is_if_let = || {
+        matches!(
+            tcx.hir().parent_iter(id).nth(1).unwrap().1,
+            hir::Node::Expr(hir::Expr { kind: hir::ExprKind::If { .. }, .. })
+        )
+    };
     tcx.struct_span_lint_hir(IRREFUTABLE_LET_PATTERNS, id, span, |lint| match ls {
-        LetSource::GenericLet => {
-            emit_diag!(lint, "`let`", "`let` is useless", "removing `let`");
-        }
-        LetSource::IfLet => {
+        hir::LetSource::If if is_if_let() => {
```

Comment:
In which cases can we have `LetSource::If` without `is_if_let` begin true? Should LetSource handle this case?

Reply:
Basically if-let-chains. The relevant test is [here](https://github.com/rust-lang/rust/blob/47ab5f7ce27397310bd8359b8db1504fbf8a9b59/src/test/ui/pattern/issue-82290.stderr#L17).

Reply:
In that case, it would be better to have an if/else block inside the branch. `is_if_let` can be renamed to something like `single_let` with a comment about if-let-chains.